lxd-user/lxd.go

```diff
@@ -0,0 +1,255 @@
+	pw, err := user.LookupId(fmt.Sprintf("%d", uid))
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve user information: %w", err)
+	}
+
+	// Setup reverter.
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Create certificate directory.
+	err = os.MkdirAll(userPath, 0700)
+	if err != nil {
+		return fmt.Errorf("Failed to create user directory: %w", err)
+	}
+	revert.Add(func() { os.RemoveAll(userPath) })
+
+	// Generate certificate.
+	err = shared.FindOrGenCert(filepath.Join(userPath, "client.crt"), filepath.Join(userPath, "client.key"), true, false)
+	if err != nil {
+		return fmt.Errorf("Failed to generate user certificate: %w", err)
+	}
+
+	// Connect to LXD.
+	client, err := lxd.ConnectLXDUnix("", nil)
+	if err != nil {
+		return fmt.Errorf("Unable to connect to LXD: %w", err)
+	}
+
+	client.GetServer()
+
+	// Setup the project (with restrictions).
+	projects, err := client.GetProjectNames()
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve project list: %w", err)
+	}
+
+	if !shared.StringInSlice(projectName, projects) {
+		// Create the project.
+		err := client.CreateProject(api.ProjectsPost{
+			Name: projectName,
+			ProjectPut: api.ProjectPut{
+				Description: fmt.Sprintf("User restricted project for %q (%s)", pw.Username, pw.Uid),
+				Config: map[string]string{
+					"features.images":               "true",
+					"features.networks":             "false",
+					"features.profiles":             "true",
+					"features.storage.volumes":      "true",
+					"restricted":                    "true",
+					"restricted.containers.nesting": "allow",
+					"restricted.devices.disk":       "allow",
+					"restricted.devices.disk.paths": pw.HomeDir,
+					"restricted.devices.gpu":        "allow",
+					"restricted.idmap.uid":          pw.Uid,
+					"restricted.idmap.gid":          pw.Gid,
+				},
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("Unable to create project: %w", err)
+		}
+
+		revert.Add(func() { client.DeleteProject(projectName) })
+	}
+
+	// Parse the certificate.
+	x509Cert, err := shared.ReadCert(filepath.Join(userPath, "client.crt"))
+	if err != nil {
+		return fmt.Errorf("Unable to read user certificate: %w", err)
+	}
+
+	// Add the certificate to the trust store.
+	err = client.CreateCertificate(api.CertificatesPost{
+		CertificatePut: api.CertificatePut{
+			Name:        fmt.Sprintf("lxd-user-%d", uid),
+			Type:        "client",
+			Restricted:  true,
+			Projects:    []string{projectName},
+			Certificate: base64.StdEncoding.EncodeToString(x509Cert.Raw),
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("Unable to add user certificate: %w", err)
+	}
+
+	revert.Add(func() { client.DeleteCertificate(shared.CertFingerprint(x509Cert)) })
+
+	// Setup default profile.
+	err = client.UseProject(projectName).UpdateProfile("default", api.ProfilePut{
+		Description: "Default LXD profile",
+		Config: map[string]string{
+			"raw.idmap": fmt.Sprintf("uid %s %s\ngid %s %s", pw.Uid, pw.Uid, pw.Gid, pw.Gid),
```

Comment:
As this is pretty important from a security perspective it would be good to add a comment here explaining why it is the day it is.